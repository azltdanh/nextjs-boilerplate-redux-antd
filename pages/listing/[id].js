import { Row, Col, Button } from 'antd';
import { getAllListingIds, getListingDetails } from '@services/listing';

export const getStaticPaths = async () => {
  const paths = await getAllListingIds();
  return {
    paths,
    fallback: false,
  };
};

export const getStaticProps = async ({ params }) => {
  const listingItem = await getListingDetails(params.id);
  return {
    props: {
      listingItem: listingItem,
    },
  };
};

const ListingItem = ({ listingItem }) => {
  const layoutWrap = {
    xs: 24,
    md: 12,
  };

  const itemWrap = {
    span: 12,
  };

  return (
    <article>
      <img src={listingItem.images[0]} alt={listingItem.title} width={360} />
      <h1>{listingItem.title}</h1>
      <Row>
        <Button type="primary" htmlType="submit" block>
          ĐẶT XE
        </Button>
      </Row>
      <Row>
        <Col {...layoutWrap}>ĐẶC ĐIỂM</Col>
        <Col {...layoutWrap}>
          <Row>
            <Col {...itemWrap}>Số ghế: {listingItem.seats}</Col>
            <Col {...itemWrap}>Truyền động: {listingItem.type}</Col>
            <Col {...itemWrap}>Nhiên liệu: {listingItem.fuel}</Col>
          </Row>
        </Col>
      </Row>
      <Row>
        <Col {...layoutWrap}>MÔ TẢ</Col>
        <Col {...layoutWrap}>{listingItem.desc}</Col>
      </Row>
    </article>
  );
};

export default ListingItem;
