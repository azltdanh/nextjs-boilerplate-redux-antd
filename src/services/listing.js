import api from './api';

export const getAllListingIds = async () => {
  const resp = await api.get('/listing');
  return resp.data.map((item) => ({
    params: {
      id: item.id.toString(),
    },
  }));
};

export const getListing = async ({ page = 1, sort = 'id', order = 'asc' }) => {
  const resp = await api.get(
    `/listing?_page=${page}&_limit=10&_sort=${sort}&_order=${order}`
  );
  return resp.data;
};

export const getListingDetails = async (id) => {
  const resp = await api.get(`/listing/${id}`);
  return resp.data;
};
