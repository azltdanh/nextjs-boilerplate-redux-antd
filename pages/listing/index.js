import Link from 'next/link';
import { getListing } from '@services/listing';

export const getServerSideProps = async () => {
  const allListing = await getListing({});
  return {
    props: {
      allListing: allListing,
    },
  };
};

const Listing = ({ allListing }) => {
  return (
    <section>
      <h2>Listing</h2>
      <ul>
        {allListing &&
          allListing.map(({ id, title }) => (
            <li key={id}>
              <Link href={`/listing/${id}`}>
                <a>{title}</a>
              </Link>
            </li>
          ))}
      </ul>
    </section>
  );
};

export default Listing;
