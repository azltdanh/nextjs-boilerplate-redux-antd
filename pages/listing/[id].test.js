import * as listingServices from '@services/listing';
import mockListingItem from '@mock/listingItem';
import ListingItem, { getStaticPaths, getStaticProps } from './[id]';

describe(ListingItem.name, () => {
  const mockPath = { params: { id: '1' } };
  it('getStaticPaths', async () => {
    const getAllListingIdsSpy = jest
      .spyOn(listingServices, 'getAllListingIds')
      .mockImplementation(() => Promise.resolve([mockPath]));
    const staticPaths = await getStaticPaths();
    expect(getAllListingIdsSpy).toHaveBeenCalledTimes(1);
    expect(staticPaths.paths).toEqual([mockPath]);
  });

  it('getStaticProps', async () => {
    const getListingDetailsSpy = jest
      .spyOn(listingServices, 'getListingDetails')
      .mockImplementation(() => Promise.resolve(mockListingItem));
    const staticProps = await getStaticProps(mockPath);
    expect(getListingDetailsSpy).toHaveBeenCalledTimes(1);
    expect(staticProps.props.listingItem).toEqual(mockListingItem);
  });

  it('renders', () => {
    const props = {
      listingItem: { ...mockListingItem },
    };
    const wrapper = shallow(<ListingItem {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
