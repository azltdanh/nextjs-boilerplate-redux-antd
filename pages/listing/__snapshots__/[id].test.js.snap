// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingItem renders 1`] = `
<article>
  <img
    alt="KIA MORNING SI 2017"
    src="https://n1-pstg.mioto.vn/cho_thue_xe_o_to_tu_lai_thue_xe_du_lich_hochiminh/kia_morning_si_2017/p/g/2019/05/04/21/iPCzb0gnXrfcTPXA_e91gQ.jpg"
    width={360}
  />
  <h1>
    KIA MORNING SI 2017
  </h1>
  <Row>
    <Button
      block={true}
      ghost={false}
      htmlType="submit"
      loading={false}
      type="primary"
    >
      ĐẶT XE
    </Button>
  </Row>
  <Row>
    <Col
      md={12}
      xs={24}
    >
      ĐẶC ĐIỂM
    </Col>
    <Col
      md={12}
      xs={24}
    >
      <Row>
        <Col
          span={12}
        >
          Số ghế: 
          4
        </Col>
        <Col
          span={12}
        >
          Truyền động: 
          Số sàn
        </Col>
        <Col
          span={12}
        >
          Nhiên liệu: 
          Xăng
        </Col>
      </Row>
    </Col>
  </Row>
  <Row>
    <Col
      md={12}
      xs={24}
    >
      MÔ TẢ
    </Col>
    <Col
      md={12}
      xs={24}
    >
      Kia Morning Si MT, xe gia đình sạch sẽ mới 53k km, bảo dưỡng đầy đủ, đầu DVD, camera de, máy lạnh mát sâu, thích hợp cho gia đình nhỏ, xe tiết kiệm xăng.
    </Col>
  </Row>
</article>
`;
